src/test/java/com/google/devtools/build/lib/analysis/VisibilityTest.java

```diff
@@ -166,6 +166,44 @@ public void testDataVisibilityPrivateCheckPrivateAtUse() throws Exception {
     assertThrows(ViewCreationFailedException.class, () -> update("//use:world"));
   }
 
+  @Test
+  public void testConfigSettingVisibilityAlwaysCheckedAtUse() throws Exception {
+    scratch.file(
+        "BUILD",
+        "load('//build_defs:defs.bzl', 'my_rule')",
+        "my_rule(",
+        "    name = 'my_target',",
+        "    value = select({",
+        "        '//config_setting:my_setting': 'foo',",
+        "        '//conditions:default': 'bar',",
+        "    }),",
+        ")");
+    scratch.file("build_defs/BUILD");
+    scratch.file(
+        "build_defs/defs.bzl",
+        "def _my_rule_impl(ctx):",
+            "    pass",
+            "my_rule = rule(",
+            "    implementation = _my_rule_impl,",
+            "    attrs = {",
+            "        'value': attr.string(mandatory = True),",
+            "    },",
+            ")"
+    );
+    scratch.file(
+        "config_setting/BUILD",
+        "config_setting(",
+        "    name = 'my_setting',",
+        "    values = {'cpu': 'does_not_matter'},",
+        "    visibility = ['//:__pkg__'],",
```

Comment:
Does this intersect at all with https://github.com/bazelbuild/bazel/issues/12932 or https://github.com/bazelbuild/bazel/issues/12933?

Reply:
It does, the visibility check is bypassed with `--noincompatible_config_setting_private_default_visibility`, which means that this issue has only been around since the flip of `--incompatible_config_setting_private_default_visibility` pre Bazel 6.

Reply:
Wait, isn't that still off by default? https://github.com/bazelbuild/bazel/blob/895969de075d66b8c11cce7ba663ec91e392d3a3/src/main/java/com/google/devtools/build/lib/pkgcache/PackageOptions.java#L137-L138

Or is this the connection?
https://github.com/bazelbuild/bazel/blob/895969de075d66b8c11cce7ba663ec91e392d3a3/src/main/java/com/google/devtools/build/lib/pkgcache/PackageOptions.java#L123-L126

Reply:
Yes, that is the connection, the issue goes away with `--noincompatible_enforce_config_setting_visibility`. Without the explicit `visibility` on the `config_setting`, the test added in this PR passes without the changes to the production code.